clang/lib/Sema/SemaTemplateDeduction.cpp

```diff
@@ -507,10 +507,62 @@ static TemplateDeductionResult DeduceNonTypeTemplateArgument(
       S, TemplateParams, NTTP, DeducedTemplateArgument(New), T, Info, Deduced);
 }
 
+static NamedDecl *DeduceTemplateArguments(Sema &S, NamedDecl *A,
+                                          TemplateArgument Default) {
+  switch (A->getKind()) {
+  case Decl::TemplateTypeParm: {
+    auto *T = cast<TemplateTypeParmDecl>(A);
+    // FIXME: DefaultArgument can't represent a pack.
+    if (T->isParameterPack())
```

Comment:
You have an example? 
But I think you are right, however I still do not understand the fixme.

Reply:
The test `foo` in `clang/test/CXX/temp/temp.decls/temp.fct/temp.func.order/p6.cpp` exercises this, but it doesn't involve any default arguments, so any effects are incidental / harmless.

I have actually just come up with a relevant test case that shows something interesting that I wasn't aware before, this could possibly be new info for the core issue.

https://godbolt.org/z/rnbzhnqG7

This shows a change of behavior introduced by P05522, only instead of introducing ambiguity, it causes different candidates to be picked. Also affects GCC.

---

Regarding the FIXME:

So the (Type | NonType | Template) template parameter declarations store the default argument for themselves. They have a `setDefaultArgument` method for setting that. While for template template parameter declarations the `setDefaultArgument` takes a `TemplateArgument`, and that means it can use any kind of TemplateArgument as a default argument, including packs.

The same is not true for (Type | NonType) template parameter declarations, their `setDefaultArgument` only takes a (Type / Expression), which means a pack can't be passed.

This is done presumably to save space in those declarations, while a Type and an Expression are a single pointer, a `TemplateArgument` is considerably larger, and you would need at least a pointer and an integer to represent a pack.

To 'fix' this FIXME, we would have to change that, and allow those declarations to be larger in at least some circumstances.

Reply:
My question was more whether we want to deduce packs at all.

I think the new behavior make sense, in that it is consistent with https://eel.is/c++draft/temp.deduct#partial-11 (
I suspect because both candidates are valid, this tie breaker gets triggered)

We might change that in the future, but now packs can't be defaulted, so I struggle to come up to think of a case where considering the pack would change the order of partial ordering (and not trailing packs can't be deduced)

here are more tests which I think are consistent https://godbolt.org/z/5Ke3TGfK6

Reply:
I see.

The code I posted was an attempt to construct such an example, where a pack doesn't need to be defaulted as-written for the proposed changes in the rules to have an effect. Except that example hits the wall at something else first, and I haven't fully investigated that. I haven't excluded the possibility that such an example could be constructed yet.

So when I proposed that provisional wording, I didn't exclude packs from the equation. I don't see a reason to, as without excluding them, the rules stay simpler and more general and should still work.

If we want to exclude them for implementation efficiency reasons, we could, as presumably it could still work as-if.